sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -151,6 +164,30 @@ abstract class Expression extends TreeNode[Expression] {
   }
 
   /**
+   * Returns the hash for this expression. Expressions that compute the same result, even if
+   * they differ cosmetically should return the same hash.
+   */
+  def semanticHash() : Int = {
```

Comment:
Identical hash values doesn't mean the identical values, I am suggesting to use the `hashCode` plus `semanticEquals` to identity the common expression, that's also the motivation of `semanticEquals`. See `AttributeReference` for details.